RInAppMessaging/Classes/CampaignDispatcher.swift

```diff
@@ -58,41 +61,47 @@ internal class CampaignDispatcher: CampaignDispatcherType {
                 return
             }
 
-            var campaign = self.queuedCampaigns.removeFirst()
-
+            let campaign = self.queuedCampaigns.removeFirst()
             let permissionResponse = self.permissionService.checkPermission(forCampaign: campaign.data)
             if permissionResponse.performPing {
                 self.delegate?.performPing()
             }
+
             guard campaign.data.isTest || permissionResponse.display else {
                 self.dispatchNext()
                 return
             }
 
-            if let updatedCampaign = self.campaignRepository.decrementImpressionsLeftInCampaign(campaign) {
-                campaign = updatedCampaign
-            } else {
-                CommonUtility.debugPrint("""
-                    Error: Campaign (\(campaign.id)) does not exist in the repository anymore (race condition?). Proceeding with old data...
-                    """)
-            }
-
             let campaignTitle = campaign.data.messagePayload.title
             self.router.displayCampaign(campaign, confirmation: {
                 let contexts = campaign.contexts
                 guard let delegate = self.delegate, !contexts.isEmpty, !campaign.data.isTest else {
+                    self.campaignRepository.decrementImpressionsLeftInCampaign(campaign)
                     return true
                 }
-
-                return delegate.shouldShowCampaignMessage(title: campaignTitle,
-                                                          contexts: contexts)
-            }(), completion: { [weak self] in
-                guard let self = self else {
-                    return
+                let shouldShow = delegate.shouldShowCampaignMessage(title: campaignTitle,
+                                                                    contexts: contexts)
+                if shouldShow {
+                    self.campaignRepository.decrementImpressionsLeftInCampaign(campaign)
```

Comment:
Why decrement impressions here and not in the display completion closure?

Reply:
I think I put it here in order to prevent adding a campaign to the queue again when its message is being displaye and it shouldn't be displayed anymore. For example when a campaign with maxImpressions 1 is being displayed it could be still added to the queue again if some event was logged in the background and impressionsLeft wasn't decremented. Actually I'm waiting here for delegate call to return first which is already too late. I need to move it earlier (where it originally was) and add a method to increment impressionsLeft when message display was cancelled by host app.